Content.Shared/_Moffstation/Cards/Systems/PlayingCardsSystem.Transfer.cs

```diff
@@ -0,0 +1,351 @@
+    public void Add<TStack>(
+        Entity<TStack> stack,
+        IEnumerable<Entity<PlayingCardComponent>> cards,
+        EntityCoordinates cardsCoords,
+        EntityUid? user
+    ) where TStack : PlayingCardStackComponent
+    {
+        TransferImpl(
+            (range, _) => cards.Take(range).Select(it => new CardLike.Entity(it)),
+            cardsCoords,
+            CardSinkFrom(stack),
+            Transform(stack).Coordinates,
+            ..,
+            user
+        );
+    }
+
+
+    /// A source of cards for transferring. Cards in the specified <paramref name="range"/> are taken from their current
+    /// location and yielded in the returned enumerable. The <paramref name="user"/> is also provided so that things
+    /// like predicted audio can be played.
+    private delegate IEnumerable<CardLike> CardSource(Range range, EntityUid? user);
+
+    /// A sink for cards for transferring. The given <paramref name="cards"/> are added to this sink, and the first card
+    /// moved along with the total number of cards moved are returned. The first/total are used for animating the
+    /// transfer. The <paramref name="user"/> is also provided so that things like predicted audio can be played.
+    private delegate (Entity<PlayingCardComponent>? firstCard, int numTotalCards) CardSink(
+        IEnumerable<CardLike> cards,
+        EntityUid? user
+    );
+
+    /// This type represents a Card either as an entity with <see cref="PlayingCardComponent"/> or an unspawned card in
+    /// a <see cref="PlayingCardDeckComponent"/> during transfers.
+    private abstract record CardLike : ISealedInheritance
+    {
+        // Private constructor to seal inheritance.
+        private CardLike() { }
+
+        public sealed record Entity(Entity<PlayingCardComponent> Ent) : CardLike;
+
+        public sealed record Unspawned(PlayingCardInDeck Data) : CardLike;
+    }
+
+
+    /// <summary>
+    /// The implementation of all card movement into and out of stacks.
+    /// </summary>
+    /// <param name="source">The implementation of taking and returning cards from the source</param>
+    /// <param name="sourceCoords">Where the source is, conceptually. Movement animations and audio play from here</param>
+    /// <param name="sink">The implementation of adding cards to the sink</param>
+    /// <param name="sinkCoords">Where the sink is, conceptually. Movement animations and audio play from here</param>
+    /// <param name="range">Specification of which cards are taken. This is relative to <paramref name="source"/>.</param>
+    /// <param name="user">Who is causing the moving. This is used to make audio / pickup animations predictive. If this
+    /// is not specified, neither audio nor pickup animations will occur.</param>
+    private void TransferImpl(
+        CardSource source,
+        EntityCoordinates sourceCoords,
+        CardSink sink,
+        EntityCoordinates sinkCoords,
+        Range range,
+        EntityUid? user
+    )
```

Comment:
All of them delegate to this implementation because ultimately all we're doing is moving cards from a source to a sink.
`CardSource` and `CardSink` are both delegate types which more or less just make and consume an `Enumerable` of cards that're moved from the source to the sink.